proposals/2117-marquez-over-time.md

```diff
@@ -0,0 +1,29 @@
+# Proposal: Marquez API to retrieve historical data
+
+## Problem
+
+Marquez data model in PostgresSQL allows extracting a snapshot of any lineage content from the past. This may be extremely
+useful to track how job, dataset etc. evolved over time. The feature can be implemented within each Marquez endpoint: like
+retrieving datasets or jobs. The purpose of this doc is to agree on a consistent API implementation approach. 
+
+## Solution
+
+Point in time can be described by a datetime, a dataset version, a job version or even by runId. We decide that versions are the
+only reliable parameter for a point in time lineage API. Datetime can be ambiguous, for instance by having START and COMPLETE
+datetime for each job.
+
+We assume the ability to scan over time can be implemented within UI. 
+
+We introduce `snapshotAt` URL parametere and provide below some example values of it:
+
+ * `/api/v1/namespaces/some-namespace/datasets/some-dataset?snapshotAt=dataset_version:5ca3b37e-4e18-11ed-bdc3-0242ac120002`
+ * `/api/v1/namespaces/some-namespace/jobs/some-job?snapshotAt=job_version:5ca3b37e-4e18-11ed-bdc3-0242ac120002`
+ * `/api/v1/namespaces/some-namespace/jobs/some-job?snapshotAt=run_id:5ca3b37e-4e18-11ed-bdc3-0242ac120002`
```

Comment:
How are these different from the `/versions` variant of the APIs? 

https://github.com/MarquezProject/marquez/blob/main/api/src/main/java/marquez/api/JobResource.java#L111-L131
https://github.com/MarquezProject/marquez/blob/main/api/src/main/java/marquez/api/DatasetResource.java#L101-L121

Reply:
`versions` variant of the APIs are implemented for `Dataset` and `Job` and this approach does not seem to be extendable to `lineage` or `column-lineage` endpoints. It makes sense to ask for lineage at specific `run_id` or lineage of a specific `dataset_version`. Lineage can be versioned by multiple params like tracking how the lineage looked like for different dataset versions or how did it look like for certain run_ids.

Reply:
Ok, then if this is particular to the `/lineage` and `/column-lineage` APIs, let's update the URLs here to point to those APIs rather than the `/datasets` and `/jobs` APIs. I don't think we need to change the existing job/dataset APIs, as they're already being used.

Currently, the `/lineage` and `/column-lineage` APIs accept a `NodeId` as their argument. If asking for a `run`, we shouldn't need any extra parameters, as the point-in-time parameter is inferred. For the `job` and `dataset` nodes, can we simply pass in a `version` parameter? Whether by modifying the node id (e.g., `job:abc@version` or something similar) or by passing in a query parameter.